services/src/main/java/org/keycloak/services/managers/AuthenticationSessionManager.java

```diff
@@ -242,4 +247,43 @@ public AuthenticationSessionModel getAuthenticationSessionByIdAndClient(RealmMod
         RootAuthenticationSessionModel rootAuthSession = session.authenticationSessions().getRootAuthenticationSession(realm, authSessionId);
         return rootAuthSession==null ? null : rootAuthSession.getAuthenticationSession(client, tabId);
     }
+
+    public UserSessionModel getUserSessionFromAuthenticationCookie(RealmModel realm) {
+        String oldEncodedId = getAuthSessionCookies(realm);
+
+        if (oldEncodedId == null) {
+            // ideally, we should not rely on auth session id to retrieve user sessions
+            // in case the auth session was removed, we fall back to the identity cookie
+            // we are here doing the user session lookup twice, however the second lookup is going to make sure the
+            // session exists in remote caches
+            AuthenticationManager.AuthResult authResult = authenticateIdentityCookie(session, realm, true);
+
+            if (authResult != null && authResult.getSession() != null) {
+                oldEncodedId = authResult.getSession().getId();
+            } else {
+                return null;
+            }
+        }
+
+        AuthSessionId authSessionId = decodeAuthSessionId(oldEncodedId);
+        String sessionId = authSessionId.getDecodedId();
+
+        // TODO: remove this code once InfinispanUserSessionProvider is removed or no longer using any remote caches, as other implementations don't need this call.
```

Comment:
I know this TODO is from the old code, but is it still relevant? If so, can we use the feature gates to prevent this code being executed when remote caches aren't required?

Reply:
The session is cached locally in the transaction context on the first access. So, this invocation will do a remote get and cache the user session and the next invocation is going to read from memory. 
My 2 cents, it is not worth making the code more "complex".

Reply:
See 
https://github.com/keycloak/keycloak/blob/e62604b1ec64da2981b02ba2868994f61908e6a0/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/transaction/RemoteChangeLogTransaction.java#L128-L134
and
https://github.com/keycloak/keycloak/blob/e62604b1ec64da2981b02ba2868994f61908e6a0/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/transaction/RemoteChangeLogTransaction.java#L185-L192

Reply:
Thanks for explaining. Maybe we should remove the comment altogether, WDYT @ahus1 ?

Reply:
It took a while from Pedro and me to figure out what the code was trying to do here. And as that code should eventually be removed, IMHO it is valid to keep it so others won't need to put in the same amount of time again. 

I've had a look when this code is being called. I found that whenever you are already logged in and you try to access a client (for example, log in to the admin console, and then switch to the account console). Or you press refresh in your browser.

IMHO this is a common, but not too frequent case, so it wouldn't bring much of a performance improvement. Also given the comment from Pedro that the entry is fetched only once, and then re-used inside of the transaction indicates that adding a condition around wouldn't have much of a performance benefit. 

I'd say to keep it as is until we remove it.